### ISSUE TITLE
Lasers are leading targets

### ISSUE THREAD

**Issue by jafdy** [endless-sky/endless-sky#1416]
Laser AI seems to be trying to lead targets (like projectile weapons do) but this means they miss. 
This is most noticeable with turrets against circling fighter/interceptors.
I have been trying and failing to get a screen shot.

**Comment by toilethinges** [comment]
Causing things with continuous firerate to not lead wouldn't necessarily work as a fix, because of burst lasers like the Slicer.

**Comment by jafdy** [comment]
I could swear that they didn't do this in 0.8.

**Comment by Bladewood** [comment]
I think they did.
It was quite annoying when my ship kept angling to the side and only caused my plasma cannons to hit the Navy, ignoring my heavy lasers.

**Comment by KiLEdEnNis** [comment]
it started around 1.9 I think, I just think it's a targetting issues, because the lasers permanently focus a corner of the "square" the ship sprite is in

**Comment by jafdy** [comment]
You meen 0.9.1. I'm glad others are noticing this too.

**Comment by KiLEdEnNis** [comment]
I've had it all the time but I figured someone would have made an issue months ago

**Comment by Bladewood** [comment]
Unfortunately, the only viable solution to this issue which I can see is to only use one 'class' of weapons, either projectile weapons or beam weapons.

We want both of them to hit, but because of Endless Sky physics it's impossible for both to happen at the same time. We could have the superior class of weapons 'lead' the target, but right now we can't do much.

**Comment by jafdy** [comment]
Regardless it's a bug that needs fixing.

**Comment by MarcelineVQ** [comment]
Regarding fixed mount guns: I haven't seen this issue with just lasers, but I have seen what @Bladewood has reported where the ship will lead to hit with shot-based weapons causing beams to miss. I also think the fix, if there is one, is to have the lead range with the highest dps to be preferred.
This is a reality of fixed point weaponry. However if at some time gun mounts become slightly turreted this could be alleviated, and could make for a good enhancement.
So for example a fixed mount weapon can only fire forward, but the weapon in that slot might have an attribute like `swivel 15` which means that it can track up to 15 degrees away from its facing.

All that being said @jafdy is reporting this issue with turrets not fixed weapons and turrets really have no reason not to track properly since they have freedom of motion, so if it can be confirmed it should be fixed.

**Comment by endless-sky** [comment]
There are three situations where this can happen:

1. Your ship has a mix of fixed guns: some beams, and some slower
projectiles. As a result, when firing fixed guns with auto-aim activated,
your ship will try to "compromise" between the appropriate amount of
leading for each and that may result in the beams missing. (This includes
fixed secondary weapons too, like rockets or javelins.) In this case, your
ship is doing the best it can with an impossible situation - there's no
single angle that would allow all its weapons to hit.

2. To make weapon fire look more organic and less inhumanly accurate, when
turrets first open fire they will not be aimed quite perfectly; the
inaccuracy decreases as you continue to fire and the gunners are able to
"lock on" to the target better. If this is your problem, the beams should
stop missing after firing for a second or so.

3. If what you're seeing is neither of the above issues, it could be an
actual bug. But even with the fastest, smallest fighters in the game (the
Korath drones) my experience has been that beam weapons lock onto them
rather quickly.

**Comment by jafdy** [comment]
A while ago i saw a blackbird with 2 heavy laser turrets firing at a sparrow that was flying around it. it spent 5 or 6 seconds missing before I destroyed the sparrow. I think the "inhumanly accurate" hitting is better than this humanly inaccurate missing for seconds on end.

**Comment by endless-sky** [comment]
That would have been a merchant. Small merchant ships have four times the
inaccuracy that the player has, to simulate the fact that they really don't
know how to fight.

Have you also seen this happen with the player's ship, or not?

**Comment by KiLEdEnNis** [comment]
I have the issue a lot (mostly small ships), with beam weapons only (gun AND turret), so if it can't be fixed I guess the inaccuracy should be removed :(

**Comment by endless-sky** [comment]
@KiLEdEnNis, what kind of weapons are you using? (And, are you using automatic firing?) Even testing against Korath drones my beams hardly ever miss.

**Comment by jafdy** [comment]
Yes that one was a merchant, it was just an excellent example of it. I have also seen the FW, militia, Navy, and Pirates do it and I would expect some other factions would do as well.
Even if this was restricted to merchants then it still needs looking at. (my opinion)

**Comment by KiLEdEnNis** [comment]
@endless-sky usually electron beams, but I remember having that with heavy lasers and even skylances, but yes I have automatic firing on

**Comment by endless-sky** [comment]
I ran some simulations on the previous way that ships' aim inaccuracy drifted, and found that there were some cases where the aim would continue to be off for a long time. So, I switched to a different way of simulating inaccuracy, which tries to better simulate a person adjusting their aim to track a target (0ab7ec0):

![aim](https://cloud.githubusercontent.com/assets/11450262/17626071/ad1a13b8-6079-11e6-9a74-cdcd5643ffe0.png)

Once they've been firing for a few seconds, even a merchant will "lock on" well enough to never miss by more than 10-15 pixels. But more importantly, unlike the old method, if they do miss they'll quickly correct and move back toward the proper aim instead of sticking with a bad aim for a long time.
